docs/lang/control-and-nondeterminism.md

```diff
@@ -0,0 +1,509 @@
+is not a transition that is accepted by `Next`.
+
+So far, we only considered unconditional operators. Let's have a look at the
+operator `A`:
+
+```tla
+A ==
+  y > x /\ y' = x /\ x' = x
+```
+
+If we evaluate `A` in the partial state `[x |-> 3, y |-> 10, x' |-> Null, y'
+|-> Null]`, we produce the state `[x |-> 3, y |-> 10, x' |-> 3, y' |-> 3]`.
+However, if we evaluate `A` in the partial state `[x |-> 10, y |-> 3, x' |->
+Null, y' |-> Null]`, the leftmost condition `y > x` fails, and thus there is
+no next state from that partial state that would be accepted by `A`.
+
+Until this moment, we have been considering only deterministic examples, that is,
+there was no "branching" in our reasoning. Such examples can be easily put into
+a program. What about the operators, where we can choose from multiple options
+that are simultaneously enabled? We introduce an oracle to resolve this issue.
+
+**Oracles.** For multiple choices, we introduce an external device that we call
+an oracle. More formally, we assume that there is a device called `GUESS` that
+has the following properties:
+
+ 1. For a non-empty set `S`, a call `GUESS S` returns
+    some value `v \in S`.
+ 1. A call `GUESS {}` halts the evaluation.
+ 1. There are no assumptions about fairness of `GUESS`. It is free to return
+   elements in any order, produce duplicates and ignore some elements.
+
+Why do we call it a device? We cannot call it a function, as functions are
+deterministic by definition. For the same reason, it is not a TLA+
+operator. In logic, we would say that `GUESS` is simply a binary relation on
+sets and their elements, which would not be different from the membership
+relation `\in`.
+
+Why do we need `GUESS S` and cannot use `CHOOSE x \in S: TRUE` instead?
+Actually, `CHOOSE x \in S: TRUE` is *deterministic*. It is guaranteed to return
+the same value, when it is called on two equals sets: if `S = T`, then
+`(CHOOSE x \in S: TRUE) = (CHOOSE x \in T: TRUE)`. Our `GUESS S` does not have
+this guarantee. It is free to return an arbitrary element of `S` each time
+we call it.
+
+How to implement `GUESS S`? There is no general answer to this question.
+However, we know of multiple sources of non-determinism in computer science. So
+we can think of `GUESS S` as being one of the following implementations:
+
+ 1. `GUESS S` can be a remote procedure call in a distributed system.  Unless,
+ we have centralized control over the distributed system, the returned value of
+ RPC may be non-deterministic.
+
+ 1. `GUESS S` can be simply the user input. In this case, the user resolves
+ non-determinism.
+
+ 1. `GUESS S` can be controlled by an adversary, who is trying to break the
+ system.
+
+ 1. `GUESS S` can pick an element by calling a random number generator.
+ However, note that RNG is a very special way of resolving non-determinism: It
+ assumes probabilistic distribution of elements (usually, it is close to the
```

Comment:
Technically, RNG is also deterministic.

Reply:
OK, I am talking about pseudo-random generators here :-)